File: src/lib/messageDraft.ts
/**
 * messageDraft.ts — TypeScript port of scripts/prospects/message_draft.py.
 *
 * Generates a single FB Messenger draft per prospect for the dashboard
 * outreach card. Kept in lockstep with the Python module — the same voice
 * rules apply (villain-led, real-data compliment, no price in first touch,
 * "-Alex" sign-off). When you edit one, edit both. Tests for the canonical
 * version live in scripts/tests/test_prospects.py.
 *
 * The Python version is the source of truth: it powers any future batch
 * generation (digest emails, CSV exports, etc.). This port exists so the
 * Next.js dashboard can pre-fill drafts without spawning a Python subprocess.
 */

const LIGHTHOUSE_LEAD_THRESHOLD = 30;
const COMPLIMENT_MIN_RATING = 4.0;
const COMPLIMENT_MIN_REVIEWS = 10;

const HONORIFICS_AND_CREDENTIALS = new Set([
  "dr", "mr", "mrs", "ms", "miss",
  "esq", "esquire",
  "dds", "dmd", "md", "msd", "phd",
  "jr", "sr", "ii", "iii", "iv",
]);

const RECENCY_ADVERBS = [
  "this morning",
  "earlier today",
  "today",
  "this week",
  "earlier this week",
  "the other day",
] as const;

export interface DraftableProspect {
  place_id?: string | null;
  business_name?: string | null;
  vertical?: string | null;
  decision_maker_name?: string | null;
  google_rating?: number | null;
  google_review_count?: number | null;
  audit_status?: string | null;
  website_url?: string | null;
  lighthouse_mobile_score?: number | null;
  issues?: {
    viewport_missing?: boolean | null;
    no_https?: boolean | null;
    forms_unreachable?: boolean | null;
    forms_unreachable_page?: string | null;
    stale_copyright?: number | null;
    lighthouse_mobile?: number | null;
  } | null;
}

export type IssueKey =
  | "no_website"
  | "forms_unreachable"
  | "no_viewport"
  | "no_https"
  | "lighthouse_low"
  | "stale_copyright"
  | "generic";

// ── Helpers ────────────────────────────────────────────────────────────────

function firstName(name: string | null | undefined): string | null {
  if (!name) return null;
  const tokens = name.match(/[A-Za-z][A-Za-z'’\-]*/g) ?? [];
  for (const tok of tokens) {
    const bare = tok.replace(/[^a-zA-Z]/g, "").toLowerCase();
    if (HONORIFICS_AND_CREDENTIALS.has(bare)) continue;
    if (bare.length < 2) continue;
    if (tok === tok.toUpperCase()) {
      return tok.charAt(0).toUpperCase() + tok.slice(1).toLowerCase();
    }
    return tok;
  }
  return null;
}

function hostname(url: string | null | undefined): string | null {
  if (!url) return null;
  try {
    const u = new URL(url.includes("://") ? url : "https://" + url);
    let host = u.hostname.toLowerCase();
    if (host.startsWith("www.")) host = host.slice(4);
    return host || null;
  } catch {
    return null;
  }
}

/**
 * Stable hash → index. Same prospect gets the same recency phrase across
 * reruns; different prospects get different phrases so the same word doesn't
 * appear in every drafted message.
 *
 * Tiny FNV-1a — collision properties don't matter, we just need spread.
 */
function stableIndex(seed: string, mod: number): number {
  let h = 0x811c9dc5;
  for (let i = 0; i < seed.length; i++) {
    h ^= seed.charCodeAt(i);
    h = Math.imul(h, 0x01000193);
  }
  return Math.abs(h) % mod;
}

function recencyPhrase(seed: string): string {
  return RECENCY_ADVERBS[stableIndex(seed, RECENCY_ADVERBS.length)];
}

function formTargetLabel(p: DraftableProspect): string {
  const host = hostname(p.website_url) ?? "your site";
  const pageUrl = p.issues?.forms_unreachable_page;
  if (!pageUrl) return host;
  try {
    const path = new URL(pageUrl).pathname.replace(/\/$/, "");
    return path && path !== "" ? `${host}${path}` : host;
  } catch {
    return host;
  }
}

function complimentLine(p: DraftableProspect): string | null {
  const rating = p.google_rating;
  const reviews = p.google_review_count;
  if (rating == null || reviews == null) return null;
  if (rating < COMPLIMENT_MIN_RATING || reviews < COMPLIMENT_MIN_REVIEWS) return null;
  return `${rating.toFixed(1)} stars across ${reviews} Google reviews. That reputation is earned.`;
}

function greeting(name: string | null): string {
  return name ? `Hello ${name},` : "Hello,";
}

function assemble(g: string, compliment: string | null, body: string[]): string {
  const parts: string[] = [g];
  if (compliment) parts.push(compliment);
  parts.push(...body);
  parts.push("-Alex");
  return parts.join("\n\n");
}

// ── Issue picker ───────────────────────────────────────────────────────────

export function pickTopIssue(p: DraftableProspect): IssueKey {
  if (p.audit_status === "no_website" || !p.website_url) return "no_website";
  const issues = p.issues ?? {};
  if (issues.forms_unreachable) return "forms_unreachable";
  if (issues.viewport_missing) return "no_viewport";
  if (issues.no_https) return "no_https";

  const lh = p.lighthouse_mobile_score ?? issues.lighthouse_mobile ?? null;
  if (lh != null && lh < LIGHTHOUSE_LEAD_THRESHOLD) return "lighthouse_low";

  if (issues.stale_copyright) return "stale_copyright";
  return "generic";
}

// ── Templates ──────────────────────────────────────────────────────────────

type TemplateFn = (
  p: DraftableProspect,
  fn: string | null,
  cmp: string | null
) => string;

const noWebsite: TemplateFn = (p, fn, cmp) => {
  const practice = p.business_name ?? "your practice";
  return assemble(greeting(fn), cmp, [
    "About half of new dental searches start on Google. Patients click " +
      "expecting a website — yours doesn't come up. They bounce to the next " +
      "result and you never hear about it.",
    `Want me to send a one-page mock of what ${practice}'s site could look ` +
      "like? Takes a day. No obligation.",
  ]);
};

const formsUnreachable: TemplateFn = (p, fn, cmp) => {
  const target = formTargetLabel(p);
  const when = recencyPhrase(`${p.place_id ?? ""}:forms`);
  return assemble(greeting(fn), cmp, [
    `Quick thing — I tried the contact form on ${target} ${when}. Submit ` +
      "returns an error. Patients fill it out, hit send, get nothing. Most " +
      "don't try again. They call the next dentist.",
    "Want me to send a free audit writeup of what's leaking? Takes a day. " +
      "No obligation.",
  ]);
};

const noViewport: TemplateFn = (p, fn, cmp) => {
  const host = hostname(p.website_url) ?? "your site";
  const when = recencyPhrase(`${p.place_id ?? ""}:mobile`);
  return assemble(greeting(fn), cmp, [
    `Pulled ${host} up on my phone ${when}. Text is microscopic, buttons ` +
      "cut off — the site isn't built for mobile. About 70% of dental " +
      "searches happen on phones now. Most patients bounce in two seconds.",
    "Want me to send a one-page mock of the mobile redesign? Takes a day. " +
      "No obligation.",
  ]);
};

const noHttps: TemplateFn = (p, fn, cmp) => {
  const host = hostname(p.website_url) ?? "your site";
  return assemble(greeting(fn), cmp, [
    `${host} doesn't have HTTPS. Chrome shows a 'Not Secure' warning to ` +
      "every patient who visits — most click away before they read anything. " +
      "That's first-impression revenue gone before they know who you are.",
    "Want me to send a free audit writeup? Takes a day. No obligation.",
  ]);
};

const lighthouseLow: TemplateFn = (p, fn, cmp) => {
  const host = hostname(p.website_url) ?? "your site";
  const score = p.lighthouse_mobile_score ?? p.issues?.lighthouse_mobile ?? null;
  const clause = score != null ? `came back at ${score}/100` : "came back very low";
  return assemble(greeting(fn), cmp, [
    `Ran a Google PageSpeed check on ${host}. Mobile score ${clause}. ` +
      "Patients bail on a slow load around three seconds — you're losing the " +
      "click before they ever read your name.",
    "Want me to send a free audit writeup? Takes a day. No obligation.",
  ]);
};

const staleCopyright: TemplateFn = (p, fn, cmp) => {
  const host = hostname(p.website_url) ?? "your site";
  const year = p.issues?.stale_copyright;
  const clause = year ? `says ${year}` : "is years out of date";
  return assemble(greeting(fn), cmp, [
    `Quick thing — the copyright on ${host} ${clause}. Most patients won't ` +
      "notice consciously, but it nudges the same instinct as a dusty lobby. " +
      "They wonder if anyone's still home.",
    "Want me to send a free audit writeup of what else might be reading " +
      "stale? Takes a day. No obligation.",
  ]);
};

const genericTemplate: TemplateFn = (p, fn, cmp) => {
  const practice = p.business_name ?? "your practice";
  return assemble(greeting(fn), cmp, [
    `Took a look at ${practice}'s site. A few things on it are quietly ` +
      "leaking inquiries — small mechanical issues most owners never see " +
      "because the site looks fine when they visit it on their own laptop.",
    "Want me to send a free audit writeup? Takes a day. No obligation.",
  ]);
};

const TEMPLATES: Record<IssueKey, TemplateFn> = {
  no_website: noWebsite,
  forms_unreachable: formsUnreachable,
  no_viewport: noViewport,
  no_https: noHttps,
  lighthouse_low: lighthouseLow,
  stale_copyright: staleCopyright,
  generic: genericTemplate,
};

/**
 * Build the FB Messenger draft string for a prospect. See module docstring
 * for voice rules. Returns "" only if the prospect dict is empty/invalid.
 */
export function generateFbMessage(p: DraftableProspect): string {
  if (!p || typeof p !== "object") return "";
  const fn = firstName(p.decision_maker_name);
  const cmp = complimentLine(p);

  // Outbound voice is dental-specific (memory note dated 2026-04-29). Other
  // verticals fall back to the vertical-neutral template.
  if (p.vertical && p.vertical !== "dental") {
    return genericTemplate(p, fn, cmp);
  }
  const key = pickTopIssue(p);
  return (TEMPLATES[key] ?? genericTemplate)(p, fn, cmp);
}

/**
 * URL we send the user to when we don't have a resolved Facebook Page —
 * pre-fills FB's pages search with the business name and city. Requires the
 * user to be logged in (they are; outbound runs from their personal FB).
 */
export function fbSearchUrl(p: { business_name?: string | null; city?: string | null }): string {
  const q = [p.business_name, p.city].filter(Boolean).join(" ");
  return `https://www.facebook.com/search/pages?q=${encodeURIComponent(q)}`;
}

File: scripts/_archive/prospects/message_draft.py
"""
message_draft.py — Generate a cold-outreach FB Messenger draft per prospect.

Pure functions over a website_prospects row dict. No DB writes, no network.
The dashboard pre-fills the output into prospect cards for one-click
copy → paste into Facebook Messenger.

Voice rules (mirrors scripts/voice_anchors.py + the user's
memory/cold_email_no_price_first_touch.md note):

  - First touch leads with the *villain* — hidden patient/revenue loss.
    The technical audit finding is the *mechanism* of the leak, never the
    headline.
  - Compliment opener uses real first-party Google review data
    (rating ≥ 4.0 AND review_count ≥ 10). Below that threshold the
    compliment is dropped — fake praise reads as agency-speak.
  - NO price, NO cadence, NO retainer terms in the first touch. The free
    one-page mock or audit writeup is the lead. Price re-enters in
    follow-up #2 only.
  - Period-separated, blunt voice. One anti-upsell beat
    ("No obligation.") per draft.
  - Sign-off "-Alex".

Templates are vertical-aware: dental prospects get dental-specific stats
("about half of new dental searches start on Google"), other verticals fall
back to a vertical-neutral generic template.
"""

from __future__ import annotations

import hashlib
import re
from typing import Optional
from urllib.parse import urlparse


# ── Thresholds ──────────────────────────────────────────────────────────────

# Below this Lighthouse mobile score the slow-load storyline takes the lead.
LIGHTHOUSE_LEAD_THRESHOLD = 30

# Compliment opener is only used when both gates clear — fake praise on a
# 3.2-star, 4-review practice torches credibility instantly.
COMPLIMENT_MIN_RATING = 4.0
COMPLIMENT_MIN_REVIEWS = 10


# ── Name + URL helpers ──────────────────────────────────────────────────────

# Honorifics + post-nominal credentials that should never become the greeting
# first name. "Dr. Jane Smith DDS" → "Jane".
_HONORIFICS_AND_CREDENTIALS = frozenset({
    "dr", "mr", "mrs", "ms", "miss",
    "esq", "esquire",
    "dds", "dmd", "md", "msd", "phd",
    "jr", "sr", "ii", "iii", "iv",
})


def _first_name(decision_maker_name: Optional[str]) -> Optional[str]:
    """Best-effort first name from a decision-maker string. Drops honorifics
    and post-nominal credentials. Preserves apostrophes/hyphens within the
    token; title-cases ALL-CAPS input. Returns None if no usable token."""
    if not decision_maker_name:
        return None
    tokens = re.findall(r"[A-Za-z][A-Za-z'’\-]*", decision_maker_name)
    for tok in tokens:
        bare = re.sub(r"[^a-zA-Z]", "", tok).lower()
        if bare in _HONORIFICS_AND_CREDENTIALS:
            continue
        if len(bare) < 2:
            continue
        if tok.isupper():
            return tok.capitalize()
        return tok
    return None


def _hostname(url: Optional[str]) -> Optional[str]:
    """Bare hostname (no scheme, no www., no path). None on parse failure."""
    if not url:
        return None
    try:
        parsed = urlparse(url if "://" in url else "https://" + url)
        host = (parsed.netloc or "").lower()
    except Exception:
        return None
    if not host:
        return None
    if host.startswith("www."):
        host = host[4:]
    return host or None


def _form_target_label(prospect: dict) -> str:
    """Return ``host`` or ``host/path`` when the broken-form page URL is
    known. Lets us cite the exact page so the prospect can verify the leak
    themselves without needing technical context."""
    host = _hostname(prospect.get("website_url")) or "your site"
    issues = prospect.get("issues") or {}
    page_url = issues.get("forms_unreachable_page")
    if not page_url:
        return host
    try:
        path = urlparse(page_url).path.rstrip("/")
    except Exception:
        return host
    if path and path != "":
        return f"{host}{path}"
    return host


def _stable_choice(seed: str, options: tuple) -> str:
    """Stable hash → index. Same prospect always gets the same variant
    across reruns; different prospects get different variants so the same
    word doesn't appear in every drafted message."""
    if not options:
        raise ValueError("options must be non-empty")
    h = hashlib.md5((seed or "").encode("utf-8")).hexdigest()
    return options[int(h[:8], 16) % len(options)]


# ── Issue picker ────────────────────────────────────────────────────────────

# Which storyline drives the draft. Severity-of-impact first; among ties,
# whichever the prospect can verify themselves wins (forms > viewport > https).
_ISSUE_KEYS = (
    "no_website",
    "forms_unreachable",
    "no_viewport",
    "no_https",
    "lighthouse_low",
    "stale_copyright",
    "generic",
)


def pick_top_issue(prospect: dict) -> str:
    """Choose the message storyline. Returns one of ``_ISSUE_KEYS``."""
    if prospect.get("audit_status") == "no_website" or not prospect.get("website_url"):
        return "no_website"

    issues = prospect.get("issues") or {}
    if issues.get("forms_unreachable"):
        return "forms_unreachable"
    if issues.get("viewport_missing"):
        return "no_viewport"
    if issues.get("no_https"):
        return "no_https"

    lh = prospect.get("lighthouse_mobile_score")
    if lh is None:
        lh = issues.get("lighthouse_mobile")
    if lh is not None and lh < LIGHTHOUSE_LEAD_THRESHOLD:
        return "lighthouse_low"

    if issues.get("stale_copyright"):
        return "stale_copyright"

    return "generic"


# ── Compliment opener ───────────────────────────────────────────────────────

def _compliment_line(prospect: dict) -> Optional[str]:
    rating = prospect.get("google_rating")
    reviews = prospect.get("google_review_count")
    if rating is None or reviews is None:
        return None
    try:
        rating_f = float(rating)
        reviews_i = int(reviews)
    except (TypeError, ValueError):
        return None
    if rating_f < COMPLIMENT_MIN_RATING or reviews_i < COMPLIMENT_MIN_REVIEWS:
        return None
    return f"{rating_f:.1f} stars across {reviews_i} Google reviews. That reputation is earned."


# ── Recency adverb (deterministic per prospect, varied across the list) ─────

_RECENCY_ADVERBS = (
    "this morning",
    "earlier today",
    "today",
    "this week",
    "earlier this week",
    "the other day",
)


def _recency_phrase(seed: str) -> str:
    return _stable_choice(seed, _RECENCY_ADVERBS)


# ── Templates ───────────────────────────────────────────────────────────────

def _greeting(first_name: Optional[str]) -> str:
    return f"Hello {first_name}," if first_name else "Hello,"


def _assemble(greeting: str, compliment: Optional[str], body_paragraphs: list[str]) -> str:
    """Stitch greeting → compliment → body paragraphs → '-Alex' with blank
    lines between each block. Body paragraphs are joined with blank lines."""
    parts: list[str] = [greeting]
    if compliment:
        parts.append(compliment)
    parts.extend(body_paragraphs)
    parts.append("-Alex")
    return "\n\n".join(parts)


def _no_website(prospect: dict, first_name: Optional[str], compliment: Optional[str]) -> str:
    practice = prospect.get("business_name", "your practice")
    return _assemble(_greeting(first_name), compliment, [
        "About half of new dental searches start on Google. Patients click "
        "expecting a website — yours doesn't come up. They bounce to the next "
        "result and you never hear about it.",
        f"Want me to send a one-page mock of what {practice}'s site could look "
        "like? Takes a day. No obligation.",
    ])


def _forms_unreachable(prospect: dict, first_name: Optional[str], compliment: Optional[str]) -> str:
    target = _form_target_label(prospect)
    when = _recency_phrase(str(prospect.get("place_id", "")) + ":forms")
    return _assemble(_greeting(first_name), compliment, [
        f"Quick thing — I tried the contact form on {target} {when}. Submit "
        "returns an error. Patients fill it out, hit send, get nothing. Most "
        "don't try again. They call the next dentist.",
        "Want me to send a free audit writeup of what's leaking? Takes a day. "
        "No obligation.",
    ])


def _no_viewport(prospect: dict, first_name: Optional[str], compliment: Optional[str]) -> str:
    host = _hostname(prospect.get("website_url")) or "your site"
    when = _recency_phrase(str(prospect.get("place_id", "")) + ":mobile")
    return _assemble(_greeting(first_name), compliment, [
        f"Pulled {host} up on my phone {when}. Text is microscopic, buttons "
        "cut off — the site isn't built for mobile. About 70% of dental "
        "searches happen on phones now. Most patients bounce in two seconds.",
        "Want me to send a one-page mock of the mobile redesign? Takes a day. "
        "No obligation.",
    ])


def _no_https(prospect: dict, first_name: Optional[str], compliment: Optional[str]) -> str:
    host = _hostname(prospect.get("website_url")) or "your site"
    return _assemble(_greeting(first_name), compliment, [
        f"{host} doesn't have HTTPS. Chrome shows a 'Not Secure' warning to "
        "every patient who visits — most click away before they read anything. "
        "That's first-impression revenue gone before they know who you are.",
        "Want me to send a free audit writeup? Takes a day. No obligation.",
    ])


def _lighthouse_low(prospect: dict, first_name: Optional[str], compliment: Optional[str]) -> str:
    host = _hostname(prospect.get("website_url")) or "your site"
    score = prospect.get("lighthouse_mobile_score")
    if score is None:
        score = (prospect.get("issues") or {}).get("lighthouse_mobile")
    score_clause = f"came back at {score}/100" if score is not None else "came back very low"
    return _assemble(_greeting(first_name), compliment, [
        f"Ran a Google PageSpeed check on {host}. Mobile score {score_clause}. "
        "Patients bail on a slow load around three seconds — you're losing the "
        "click before they ever read your name.",
        "Want me to send a free audit writeup? Takes a day. No obligation.",
    ])


def _stale_copyright(prospect: dict, first_name: Optional[str], compliment: Optional[str]) -> str:
    host = _hostname(prospect.get("website_url")) or "your site"
    year = (prospect.get("issues") or {}).get("stale_copyright")
    year_clause = f"says {year}" if year else "is years out of date"
    return _assemble(_greeting(first_name), compliment, [
        f"Quick thing — the copyright on {host} {year_clause}. Most patients "
        "won't notice consciously, but it nudges the same instinct as a dusty "
        "lobby. They wonder if anyone's still home.",
        "Want me to send a free audit writeup of what else might be reading "
        "stale? Takes a day. No obligation.",
    ])


def _generic(prospect: dict, first_name: Optional[str], compliment: Optional[str]) -> str:
    """Vertical-neutral fallback for non-dental prospects or rows where no
    individual finding clearly dominates. Uses 'inquiries' instead of
    'patients' so it reads cleanly for personal-injury too."""
    practice = prospect.get("business_name", "your practice")
    return _assemble(_greeting(first_name), compliment, [
        f"Took a look at {practice}'s site. A few things on it are quietly "
        "leaking inquiries — small mechanical issues most owners never see "
        "because the site looks fine when they visit it on their own laptop.",
        "Want me to send a free audit writeup? Takes a day. No obligation.",
    ])


_TEMPLATE_BY_ISSUE = {
    "no_website": _no_website,
    "forms_unreachable": _forms_unreachable,
    "no_viewport": _no_viewport,
    "no_https": _no_https,
    "lighthouse_low": _lighthouse_low,
    "stale_copyright": _stale_copyright,
    "generic": _generic,
}


def generate_fb_message(prospect: dict) -> str:
    """
    Build one cold-outreach DM draft for the given prospect row.

    Reads (all optional except business_name in practice — missing fields just
    drop the corresponding personalization):

        business_name, decision_maker_name, google_rating, google_review_count,
        audit_status, website_url, vertical, issues (jsonb dict),
        lighthouse_mobile_score, place_id

    Returns a multi-line string ready to copy/paste into FB Messenger.
    Raises TypeError if `prospect` is not a dict.
    """
    if not isinstance(prospect, dict):
        raise TypeError("prospect must be a dict")

    first_name = _first_name(prospect.get("decision_maker_name"))
    compliment = _compliment_line(prospect)

    # Outbound voice is dental-specific (memory: outbound is dental-only as of
    # 2026-04-29). For other verticals, use the vertical-neutral fallback.
    vertical = prospect.get("vertical")
    if vertical and vertical != "dental":
        return _generic(prospect, first_name, compliment)

    issue_key = pick_top_issue(prospect)
    fn = _TEMPLATE_BY_ISSUE.get(issue_key, _generic)
    return fn(prospect, first_name, compliment)
